### ISSUE TITLE
Getting nan as metric loss for validation set on epoch 3

### ISSUE THREAD

**Issue by exodus-12** [ben-davidson-6/Gated-SCNN#8]
Hello,

Please i keep getting nan as metric loss and the val_accuracy jumps to 99% on epoch 3 and then on epoch 4 i get 99% for train accuracy and nan for metric loss 

The nan keeps appearing when it tries to apply the gradients after accumulating weights for 4 iterations but don't understand why its nan and the accuracy jumps to 99% from 65%

**Comment by ben-davidson-6** [comment]
What version of the code are you using, and what data are you training on? There were some bugs in the original implementation if there were no gradients in your image, for example if you had a blank image in your validation set this could cause a problem.

If your not using the most recent, I'd advise updating the code to use the most recent version before trying anything else.

**Comment by exodus-12** [comment]
yeah i updated it, using MS COCO dataset and there are no blank images in the validation set

**Comment by ben-davidson-6** [comment]
Does the problem persist?

**Comment by exodus-12** [comment]
Yeah it does, i noticed the number of channels in the predicted shape_head is 1 while the number of channels in the actual edge_mask is 2 could this be the cause and when i tried viewing some results on the tensorboard the image and actual label showed but the predicted label and shape head was a blank image

**Comment by ben-davidson-6** [comment]
I think what is happening is the moving mean or variance of batch norm is going to a NaN so the blank image you are seeing is an image of NaN's. The shape head output size is fine, it gets converted to two channels using `tf.concat([1. - shape_head, shape_head], axis=-1)`. 

I don't have a good answer for this. I came across a similar issue on a few training runs but the problem stopped when I added the accumulation of gradients. When trying to debug this problem myself I came across these issues:

https://github.com/keras-team/keras/issues/11927
https://github.com/tensorflow/tensorflow/issues/21726

I think that somewhere the batch norm moving mean or variance is growing until NaN. I'm not sure how to fix it though or even how to recreate the problem it without training a whole network.

**Comment by ben-davidson-6** [comment]
in fact ignore my comment, I have found the issue:

```
python
import numpy as np
import tensorflow as tf

import os
os.environ['CUDA_VISIBLE_DEVICES'] = "1"



class bnmodel(tf.keras.layers.Layer):
    def __init__(self):
        super(bnmodel, self).__init__()
        self.bn = tf.keras.layers.BatchNormalization()

    def call(self, inputs, training=None, mask=None):
        y = self.bn(inputs, training)
        return y


model = bnmodel()

# train
# model.compile(loss='mse', optimizer=tf.keras.optimizers.Adam())

# model.summary()
for i in range(100000):
    print('Batch {}'.format(i))
    # batch size of 1 is the issue
    x = np.random.randn(1, 1, 1, 1).astype(np.float32)

    out = model(x, training=True)

    print('Maximum bn moving_mean: {}'.format(model.bn.moving_variance)) # nan <-

```

If you set your `drop_remainder=False`  in the dataset class for validation images, and make sure you have a batch_size > 1 it should work

**Comment by ben-davidson-6** [comment]
This is the relavent tensorflow issue:

https://github.com/tensorflow/tensorflow/issues/34062

I will modify the code so that you cannot use a batch_size of 1 when training

**Comment by exodus-12** [comment]
Yes it was already set to false in the code `dataset = dataset.batch(self.val_batch_size, drop_remainder=False)`
 and batch_size is 10 for 1342 images in validation images

**Comment by ben-davidson-6** [comment]
what about for training images ?

**Comment by exodus-12** [comment]
Training images is True will set it to False now

**Comment by ben-davidson-6** [comment]
Great, hopefully that solves the issue.

**Comment by exodus-12** [comment]
:( I'm getting 58% for train accuracy and 2.1% for validation accuracy on first epoch

**Comment by ben-davidson-6** [comment]
Your batch size is > 1 for training as well as having `drop_remainder = False`

**Comment by exodus-12** [comment]
Yes

**Comment by ben-davidson-6** [comment]
hmmm I'm not sure then. I feel fairly confident this must of been the issue. Can you update the code and try again. Now the network will duplicate tensors if you hand it a batch size of 1 so this bug should be impossible. If the error persists then we know that it must be something else

**Comment by exodus-12** [comment]
I should update the code with batch size of 1?

**Comment by ben-davidson-6** [comment]
no update to version 1.0.2 and then you can do whatever you like for the batch sizes the network shouldmake sure this case doesnt happen anymore now at the input it does

```python
        # we need to repeat the input if batch size is 1
        # because in training mode a batch size of 1 will create
        # nans, see:
        # https://github.com/tensorflow/tensorflow/issues/34062
        one_item_batch = tf.shape(inputs)[0] == 1
        if training is None:
            training = True
        inputs = tf.cond(
            tf.logical_and(one_item_batch, training),
            lambda: tf.tile(inputs, (2, 1, 1, 1)),
            lambda: inputs)
```

**Comment by exodus-12** [comment]
okay let me update it

**Comment by exodus-12** [comment]
Thanks, created my own training loop and it works now but the loss is not decreasing and the validation accuracy is stuck while the tranning accuracy changes within 3% increase or decrease acroos epochs. i have tried using adam optimizer but still getting the same thing, please any advice

**Comment by benDavidson6** [comment]
try lowering or increasing the learning rate by factors of 10 and see if it helps, the lr set for cityscapes might not be v good for coco
